File: frontend/lib/models/detection.dart
/// Detection bounding box
class DetectionBox {
  final int classId;
  final String className;
  final double confidence;
  final double x1;
  final double y1;
  final double x2;
  final double y2;
  final double centerX;
  final double centerY;
  final int? trackId;

  DetectionBox({
    required this.classId,
    required this.className,
    required this.confidence,
    required this.x1,
    required this.y1,
    required this.x2,
    required this.y2,
    required this.centerX,
    required this.centerY,
    this.trackId,
  });

  factory DetectionBox.fromJson(Map<String, dynamic> json) {
    return DetectionBox(
      classId: json['class_id'],
      className: json['class_name'],
      confidence: (json['confidence'] as num).toDouble(),
      x1: (json['x1'] as num).toDouble(),
      y1: (json['y1'] as num).toDouble(),
      x2: (json['x2'] as num).toDouble(),
      y2: (json['y2'] as num).toDouble(),
      centerX: (json['center_x'] as num).toDouble(),
      centerY: (json['center_y'] as num).toDouble(),
      trackId: json['track_id'],
    );
  }

  double get width => x2 - x1;
  double get height => y2 - y1;
}

/// Detection result for a frame
class DetectionResult {
  final int frameNumber;
  final double timestamp;
  final List<DetectionBox> detections;
  final int personsCount;
  final int helmetsCount;
  final int masksCount;
  final int fireExtinguishersCount;

  DetectionResult({
    required this.frameNumber,
    required this.timestamp,
    required this.detections,
    this.personsCount = 0,
    this.helmetsCount = 0,
    this.masksCount = 0,
    this.fireExtinguishersCount = 0,
  });

  factory DetectionResult.fromJson(Map<String, dynamic> json) {
    return DetectionResult(
      frameNumber: json['frame_number'],
      timestamp: (json['timestamp'] as num).toDouble(),
      detections: (json['detections'] as List)
          .map((d) => DetectionBox.fromJson(d))
          .toList(),
      personsCount: json['persons_count'] ?? 0,
      helmetsCount: json['helmets_count'] ?? 0,
      masksCount: json['masks_count'] ?? 0,
      fireExtinguishersCount: json['fire_extinguishers_count'] ?? 0,
    );
  }
}

/// Stream frame data from WebSocket
class StreamFrame {
  final int cameraId;
  final String frameBase64;
  final double currentMs;
  final double totalMs;
  final DetectionResult? detection;
  final List<Map<String, dynamic>> events;
  final List<Map<String, dynamic>> rois;
  final Map<String, dynamic> roiMetrics;

  StreamFrame({
    required this.cameraId,
    required this.frameBase64,
    this.currentMs = 0.0,
    this.totalMs = 0.0,
    this.detection,
    this.events = const [],
    this.rois = const [],
    this.roiMetrics = const {},
  });

  factory StreamFrame.fromJson(Map<String, dynamic> json) {
    return StreamFrame(
      cameraId: json['camera_id'],
      frameBase64: json['frame'],
      currentMs: (json['current_ms'] as num?)?.toDouble() ?? 0.0,
      totalMs: (json['total_ms'] as num?)?.toDouble() ?? 0.0,
      detection: json['detection'] != null
          ? DetectionResult.fromJson(json['detection'])
          : null,
      events: json['events'] != null
          ? List<Map<String, dynamic>>.from(json['events'])
          : [],
      rois: json['rois'] != null
          ? List<Map<String, dynamic>>.from(json['rois'])
          : [],
      roiMetrics: json['roi_metrics'] != null
          ? Map<String, dynamic>.from(json['roi_metrics'])
          : {},
    );
  }
}

File: backend/app/schemas/detection.py
"""
Detection schemas for YOLO results and streaming.
"""
from typing import List, Optional
from pydantic import BaseModel, Field


class DetectionBox(BaseModel):
    """Single detection bounding box."""
    class_id: int = Field(..., description="Class index")
    class_name: str = Field(..., description="Class name")
    confidence: float = Field(..., ge=0, le=1, description="Detection confidence")
    x1: float = Field(..., description="Top-left X coordinate")
    y1: float = Field(..., description="Top-left Y coordinate")
    x2: float = Field(..., description="Bottom-right X coordinate")
    y2: float = Field(..., description="Bottom-right Y coordinate")
    center_x: float = Field(..., description="Center X coordinate")
    center_y: float = Field(..., description="Center Y coordinate")
    track_id: Optional[int] = Field(None, description="Unique tracking ID for the object")


class DetectionResult(BaseModel):
    """Detection result for a single frame."""
    frame_number: int
    timestamp: float
    detections: List[DetectionBox]
    persons_count: int = 0
    helmets_count: int = 0
    masks_count: int = 0
    fire_extinguishers_count: int = 0


class StreamFrame(BaseModel):
    """WebSocket stream frame data."""
    camera_id: int
    frame_base64: str = Field(..., description="Base64 encoded JPEG frame")
    current_ms: float = 0.0
    total_ms: float = 0.0
    detection: Optional[DetectionResult] = None
    events: List[dict] = Field(default_factory=list, description="New events for this frame")


class SafetyStatus(BaseModel):
    """Current safety status for a camera."""
    camera_id: int
    roi_intrusions: List[dict] = Field(default_factory=list)
    ppe_violations: List[dict] = Field(default_factory=list)
    fire_extinguisher_status: bool = Field(default=False, description="Fire extinguisher detected")
    overall_status: str = Field(default="OK", description="OK, WARNING, CRITICAL")
